File: en/codes/rust/chapter_divide_and_conquer/hanota.rs
/*
 * File: hanota.rs
 * Created Time: 2023-07-15
 * Author: codingonion (coderonion@gmail.com)
 */

#![allow(non_snake_case)]

/* Move a disk */
fn move_pan(src: &mut Vec<i32>, tar: &mut Vec<i32>) {
    // Take out a disk from the top of src
    let pan = src.pop().unwrap();
    // Place the disk on top of tar
    tar.push(pan);
}

/* Solve the Tower of Hanoi problem f(i) */
fn dfs(i: i32, src: &mut Vec<i32>, buf: &mut Vec<i32>, tar: &mut Vec<i32>) {
    // If there is only one disk left in src, move it directly to tar
    if i == 1 {
        move_pan(src, tar);
        return;
    }
    // Subproblem f(i-1): move the top i-1 disks from src to buf using tar
    dfs(i - 1, src, tar, buf);
    // Subproblem f(1): move the remaining disk from src to tar
    move_pan(src, tar);
    // Subproblem f(i-1): move the top i-1 disks from buf to tar using src
    dfs(i - 1, buf, src, tar);
}

/* Solve the Tower of Hanoi problem */
fn solve_hanota(A: &mut Vec<i32>, B: &mut Vec<i32>, C: &mut Vec<i32>) {
    let n = A.len() as i32;
    // Move the top n disks from A to C using B
    dfs(n, A, B, C);
}

/* Driver Code */
pub fn main() {
    let mut A = vec![5, 4, 3, 2, 1];
    let mut B = Vec::new();
    let mut C = Vec::new();
    println!("In initial state:");
    println!("A = {:?}", A);
    println!("B = {:?}", B);
    println!("C = {:?}", C);

    solve_hanota(&mut A, &mut B, &mut C);

    println!("After disk movement is complete:");
    println!("A = {:?}", A);
    println!("B = {:?}", B);
    println!("C = {:?}", C);
}

File: en/codes/ruby/chapter_divide_and_conquer/hanota.rb
=begin
File: hanota.rb
Created Time: 2024-05-13
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

### Move one disk ###
def move(src, tar)
  # Take out a disk from the top of src
  pan = src.pop
  # Place the disk on top of tar
  tar << pan
end

### Solve Tower of Hanoi f(i) ###
def dfs(i, src, buf, tar)
  # If there is only one disk left in src, move it directly to tar
  if i == 1
    move(src, tar)
    return
  end

  # Subproblem f(i-1): move the top i-1 disks from src to buf using tar
  dfs(i - 1, src, tar, buf)
  # Subproblem f(1): move the remaining disk from src to tar
  move(src, tar)
  # Subproblem f(i-1): move the top i-1 disks from buf to tar using src
  dfs(i - 1, buf, src, tar)
end

### Solve Tower of Hanoi ###
def solve_hanota(_A, _B, _C)
  n = _A.length
  # Move the top n disks from A to C using B
  dfs(n, _A, _B, _C)
end

### Driver Code ###
if __FILE__ == $0
  # The tail of the list is the top of the rod
  A = [5, 4, 3, 2, 1]
  B = []
  C = []
  puts "In initial state:"
  puts "A = #{A}"
  puts "B = #{B}"
  puts "C = #{C}"

  solve_hanota(A, B, C)

  puts "After disk movement is complete:"
  puts "A = #{A}"
  puts "B = #{B}"
  puts "C = #{C}"
end
